### PR TITLE
feat(internal): Add gather timeout metric

### LINKED ISSUE #13470: Add a counter for collections that take longer than expected
### Use Case

When a script under an exec input simply locks up and does not produce output within the expected time, there is no internal metric to indicate a failure. There is output in the logs however.

While my issue is for an exec script, this issue could also occur for any other plugin type.

### Expected behavior

A new metric such as `timeout_warnings` under `internal_agent` is incremented.

### Actual behavior

```
# Telegraf log snippet
2023-06-20T01:14:15Z W! [inputs.exec] Collection took longer than expected; not complete after interval of 4m0s

# Internal metric snippet - No gather errors
internal_agent,agent_version=1.24.3,go_version=1.19.3,host=TEST-HOST1 metrics_written=1462052i,metrics_dropped=0i,metrics_gathered=731034i,gather_errors=0i 1687223760000000000`

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by mcassaniti**
# Required for all PRs

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #13470

Added field `timeout_warnings` to measurement `internal_agent` to count warnings generated when collection timeouts occur. This is highly useful when exec based inputs lock up but there is no error generated.

### CODE REVIEW

**Comment by mcassaniti** [comment]
!signed-cla

File: models/running_input.go [hunk 1]
```diff
@@ -10,6 +10,7 @@ import (
 var (
 	GlobalMetricsGathered = selfstat.Register("agent", "metrics_gathered", map[string]string{})
 	GlobalGatherErrors    = selfstat.Register("agent", "gather_errors", map[string]string{})
+	GlobalTimeoutWarnings = selfstat.Register("agent", "timeout_warnings", map[string]string{})
```
Comment (srebhan) [inline models/running_input.go:13]: How about using a more speaking name, e.g. `gather_timeouts`?

**Review by srebhan — COMMENTED** [review]
Thanks for this nice contribution @mcassaniti! The only request from my side is to use a more speaking and specific name for the new stat.

File: models/running_input.go (hunk 1 above)
Comment (mcassaniti) [reply models/running_input.go:13]: Done.

File: models/running_input.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 var (
 	GlobalMetricsGathered = selfstat.Register("agent", "metrics_gathered", map[string]string{})
 	GlobalGatherErrors    = selfstat.Register("agent", "gather_errors", map[string]string{})
+	GlobalTimeoutWarnings = selfstat.Register("agent", "gather_timeouts", map[string]string{})
```
Comment (srebhan) [inline models/running_input.go:13]: Sorry for being pedantic but can you please also fix the variable names!?

Comment (mcassaniti) [reply models/running_input.go]: I've redone the variable name and the associated functions.

**Review by Hipska — APPROVED** [review]
I like the idea, but wouldn't it also be nice if you could have a counter per plugin as well? So you could know which plugin is having these timeouts.

**Comment by mcassaniti** [comment]
@Hipska I've made some adjustments so lets see if that covers it.

My main driver is to have some sort of indication that there are timeouts and then staff can check the agent log for further details.

**Review by Hipska — APPROVED** [review]

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks @mcassaniti for your contribution!

**Review by powersj — APPROVED** [review]

**Comment by mcassaniti** [comment]
@powersj I'm unsure of the release process so I was wondering when this feature would be included in a release.

**Comment by Hipska** [comment]
Since it is a feature and not a bug fix, it will be included in the next feature release (1.28) scheduled for [sept 11](https://calendar.google.com/calendar/u/0/embed?src=c_1ikq7u4f5c4o6mh9ep4duo3avk@group.calendar.google.com).

@srebhan another thing that would clear up for contributers when milestones would be used a bit more..

**Pull request merged**